cluster authn trust configmap: increase write interval during apiserver dueling

stlaz opened the pull request:
<!--  Thanks for sending a pull request!  Here are some tips for you:

1. If this is your first time, please read our contributor guidelines: https://git.k8s.io/community/contributors/guide/first-contribution.md#your-first-contribution and developer guide https://git.k8s.io/community/contributors/devel/development.md#development-guide
2. Please label this pull request according to what type of issue you are addressing, especially if this is a release targeted pull request. For reference on required PR/issue labels, read here:
https://git.k8s.io/community/contributors/devel/sig-release/release.md#issuepr-kind-label
3. Ensure you have added or ran the appropriate tests for your PR: https://git.k8s.io/community/contributors/devel/sig-testing/testing.md
4. If you want *faster* PR reviews, read how: https://git.k8s.io/community/contributors/guide/pull-requests.md#best-practices-for-faster-reviews
5. If the PR is unfinished, see how to mark it: https://git.k8s.io/community/contributors/guide/pull-requests.md#marking-unfinished-pull-requests
-->

#### What type of PR is this?
/kind feature

#### What this PR does / why we need it:
This PR attempts to future-proof adding new keys to the kube-system/extension-apiserver-authentication
config map that serves as a source for in-cluster certificate authentication source.

#### Which issue(s) this PR fixes:
<!--
*Automatically closes linked issue when PR is merged.
Usage: `Fixes #<issue number>`, or `Fixes (paste link of issue)`.
_If PR is about `failing-tests or flakes`, please post the related issues/tests in a comment and do not use `Fixes`_*
-->
Fixes #129335

#### Special notes for your reviewer:

#### Does this PR introduce a user-facing change?
<!--
If no, just write "NONE" in the release-note block below.
If yes, a release note is required:
Enter your extended release note in the block below. If the PR requires additional action from users switching to the new release, include the string "action required".

For more information on release notes see: https://git.k8s.io/community/contributors/guide/release-notes.md
-->
```release-note
NONE
```

#### Additional documentation e.g., KEPs (Kubernetes Enhancement Proposals), usage docs, etc.:

<!--
This section can be blank if this pull request does not require a release note.

When adding links which point to resources within git repositories, like
KEPs or supporting documentation, please reference a specific commit and avoid
linking directly to the master branch. This ensures that links reference a
specific point in time, rather than a document that may change over time.

See here for guidance on getting permanent links to files: https://help.github.com/en/articles/getting-permanent-links-to-files

Please use the following format for linking documentation:
- [KEP]: <link>
- [Usage]: <link>
- [Other doc]: <link>
-->
```docs

```

stlaz commented:
/sig auth

seans3 commented:
/remove-sig api-machinery

stlaz commented:
/hold
I naively thought that the kube-apiserver's manager name will always be "kube-apiserver" but the failing integration test proved me wrong. I need to find out how the manager name gets derived to see if the proposed solution is feasible.

stlaz commented:
/hold cancel
The known fields are now always handled irregardless of the field manager.

stlaz commented:
/test pull-kubernetes-e2e-gce
looks like a flake in port-forwarding test

stlaz commented:
/assign enj liggitt

liggitt commented on pkg/controlplane/controller/clusterauthenticationtrust/cluster_authentication_trust_controller.go:
```diff
@@ -149,8 +149,11 @@ func (c *Controller) syncConfigMap() error {
-	// keep the original to diff against later before updating
-	authConfigMap := originalAuthConfigMap.DeepCopy()
+
+	existingData, err := extractAPIServerOwnedFields(originalAuthConfigMap)
```
Hmm... I'm really iffy on using field manager for this... is the field manager name stable and reliable? Shouldn't we just tolerate / preserve all unknown keys as-is to be forward-compatible?

stlaz replied:
The writes on the CM this controller performs now explicitly set the manager name.

I wasn't sure we want to make the controller own the CM completely or whether we're ok with any unknown fields. Either way it's possible to add fields to the CM with enough privileges, with the managed fields they only have to be a bit more explicit about it.

If it's fine to keep any unknown fields, I'll remove the managed fields condition.

liggitt commented on pkg/controlplane/controller/clusterauthenticationtrust/cluster_authentication_trust_controller.go:
```diff
@@ -256,32 +271,25 @@ func getConfigMapDataFor(authenticationInfo ClusterAuthenticationInfo) (map[stri
+	var err error
+	for key, configData := range map[string][]string{
+		"requestheader-username-headers":     authenticationInfo.RequestHeaderUsernameHeaders.Value(),
+		"requestheader-uid-headers":          authenticationInfo.RequestHeaderUIDHeaders.Value(),
```
why is it safe to unconditionally set this key again? won't that duel with 1.32 API servers?

stlaz replied:
I'll remove the UIDs beta move from this PR, let's focus on handling the existing keys behavior first.

liggitt commented on pkg/controlplane/controller/clusterauthenticationtrust/cluster_authentication_trust_controller.go:
```diff
@@ -256,32 +271,25 @@ func getConfigMapDataFor(authenticationInfo ClusterAuthenticationInfo) (map[stri
+		"requestheader-extra-headers-prefix": authenticationInfo.RequestHeaderExtraHeaderPrefixes.Value(),
+		"requestheader-allowed-names":        authenticationInfo.RequestHeaderAllowedNames.Value(),
+	} {
+		if len(configData) == 0 {
+			continue
```
this looks like it is possibly changing how *all* the existing keys are serialized to omit them when zero-length ... can we avoid messing with all the existing fields in this PR? it makes it a lot harder to reason about the cross-version impact

liggitt commented on pkg/controlplane/controller/clusterauthenticationtrust/cluster_authentication_trust_controller.go:
```diff
@@ -448,6 +454,44 @@ func filterExpiredCerts(certs ...*x509.Certificate) []*x509.Certificate {
+			continue
+		}
+		managedUnstructured := map[string]interface{}{}
+		if err := json.Unmarshal(f.FieldsV1.Raw, &managedUnstructured); err != nil {
+			return nil, fmt.Errorf("failed to unmarshal managed fields: %w", err)
+		}
+
+		dataMap, found, err := unstructured.NestedMap(managedUnstructured, "f:data")
```
hmm... parsing out field manager seems significantly more complicated and fragile than tolerating / preserving unknown keys in the configmap

stlaz commented:
Removed the UID headers bump to Beta for now, and also got rid of the managedFields logic in favour of just keeping unhandled keys.

cici37 commented:
/remove-sig api-machinery

cici37 commented:
Just a reminder that please rewrite the context for `Does this PR introduce a user-facing change?` in PR description before merging the PR. Thanks

stlaz commented:
/triage accepted
we triaged it but forgot to label it :)

liggitt commented on pkg/controlplane/controller/clusterauthenticationtrust/cluster_authentication_trust_controller.go:
```diff
@@ -160,11 +158,25 @@ func (c *Controller) syncConfigMap() error {
 		return err
 	}
 
+	// If there are keys in the original map that we failed to construct from our
```
Taking a step back, this PR is trying to make instances of this controller loop in different API servers with different views of what the data *should* be coexist more gracefully, without triggering the hotloop dueling write behavior observed in https://github.com/kubernetes/kubernetes/issues/128986

**This scenario can exist for two reasons:**
1. The API servers are at different versions that construct the configmap differently.
    * Clusters are not expected to run in skewed server versions permanently, and are expected to converge (though it could take minutes-days to do so and the servers should not _break_ while in a skewed state).
2. The API servers have different configuration that produces different configmap content (e.g. adding or removing one of the configured headers, or adding or removing one of the configured CA certificates, etc).
    * Clusters are not expected to run with non-homogenous server configurations permanently, and are expected to converge (though it could take minutes-days to do so and the servers should not _break_ while in a non-homogenous state).

**There are two ways the instances can disagree about the data:**
1. They can disagree about which keys should exist (like going from zero <-> non-zero headers configured for an item, or adding a new item unknown to previous versions)
2. They can disagree about the content for a key they both thing should exist (like going from 1 <-> 2 headers for an item, or from A <-> B header for an item)

This PR currently treats these cases differently. It only handles case 1, and still duels on case 2.

**There are three things an instance could do when asked to reconcile data that does not match its own perspective of what the data should be:**
1. Leave it as-is (not reconcile its own view)
2. Merge the data
3. Overwrite it as fast and often as possible
4. Overwrite it at some non-hotloop period

The pre-existing code chooses option 2 for keys it knows contain CA bundles.

This PR chooses option 1 for keys it doesn't think should exist, and still chooses option 3 for keys it does think should exist which are not the special merged CA bundle keys.

**I think our goals should be:**
1. Don't break / destabilize servers while in the skewed version or non-homogenous config state
2. Eventually converge the data in the configmap to the data all the servers agree should be there once they all converge versions/config

This PR currently accomplishes goal 1 only in the case of added / removed keys, and does not accomplish goal 2 if there are extra spurious keys in the configmap (those never get reconciled away).

**When do we queue the configmap to the sync loop?**

https://github.com/kubernetes/kubernetes/blob/d3548f487d0272c5bc3271e474039ab8d94e25d3/pkg/controlplane/controller/clusterauthenticationtrust/cluster_authentication_trust_controller.go#L125-L137

* AddFunc - this fires on server startup, and on recreate of the configmap. It is not involved in hotloop dueling.
* UpdateFunc - this fires on any observed update of the configmap. This was the cause of the hotloop dueling.
* DeleteFunc - this fires on any observed delete of the configmap. It is not involved in hotloop dueling.

https://github.com/kubernetes/kubernetes/blob/d3548f487d0272c5bc3271e474039ab8d94e25d3/pkg/controlplane/controller/clusterauthenticationtrust/cluster_authentication_trust_controller.go#L478-L481

* Periodic loop - this fires at server start, then once a minute, forever. It is not involved in hotloop dueling.

https://github.com/kubernetes/kubernetes/blob/d3548f487d0272c5bc3271e474039ab8d94e25d3/pkg/controlplane/controller/clusterauthenticationtrust/cluster_authentication_trust_controller.go#L513-L520

* On any error inside the sync loop. This could hit a conflict and requeue in dueling cases, but it doesn't *cause* the hotloop dueling.

**Proposed alternative: just delete UpdateFunc**

What if we left all reconciliation logic as-is, and just deleted UpdateFunc to prevent hotloop dueling?

On startup, AddFunc and periodic loop startup would trigger initial reconciliation.

If the configmap was completely deleted, DeleteFunc would recreate it instantly (and then AddFunc would observe that and reconcile again)

If some other writer modified the configmap, the controller would react / reconcile within a minute when the periodic poll occurred. That turns the hotloop dueling into slow dueling where the config map in a cluster with servers with 2 different views of the data would get written ~2 times a minute.

That seems tolerable to me, is _way_ simpler to reason about, and I think accomplishes our goals of eventually converging the data once all servers agree.

stlaz replied:
I agree the `UpdateFunc` does not appear to play a significant role and removing it to make sure we don't duel in the future seems reasonable and safe.

> The pre-existing code chooses option 2 for keys it knows contain CA bundles (merging bundles and dropping expired certs).

This does not matter much, but I think the preexisting code merges both the CAs and the values of the requestheader keys in `func combinedClusterAuthenticationInfo(lhs, rhs ClusterAuthenticationInfo) (ClusterAuthenticationInfo, error)`, and so the resulting slice values are always a union of {\<previous values\>}, {\<config values\>}. I think there are only two scenarios when keys/values would be deleted:
1. the config map gets too big, the whole CM gets deleted
2. all of requestheader CA certs expire - all of requestheader related keys get removed

Since in the case of 2. we would have a behavior change with the currently proposed changes (I only just realized this, actually), we should probably still pursue the removal of the `UpdateFunc` and keeping the original logic.

stlaz committed:
authn-trust: don't react to target CM updates

If kube-apiservers run at a different version (during upgrades) and
each has different opinion on what the `extension-apiserver-authentication`
config map should look like, they would start dueling with config map
writes.

This commit removes handling of the update events of the target CM
in order to reduce the amount of such dueling to once a minute, as
the controller is guaranteed to run at least once a minute anyway.

The idea is that a cluster state with two different versions of
kube-apiserver should never be permanent, and so a reduced amount of
dueling for that period is tolerable.

stlaz commented:
I removed the `UpdateFunc` and added timeouts into appropriate places in the integration tests. I kept the small noop refactor I made for this PR originally.

liggitt commented on test/integration/apiserver/certreload/certreload_test.go:
```diff
@@ -184,12 +184,12 @@ func testClientCA(t *testing.T, recreate bool) {
 	defer tearDownFn()
 
 	// wait for request header info
-	err = wait.PollImmediate(100*time.Millisecond, 30*time.Second, waitForConfigMapCAContent(t, kubeClient, "requestheader-client-ca-file", "-----BEGIN CERTIFICATE-----", 1))
```
I'm surprised these timeouts had to change... the change triggers here were file changes, which we're not touching, right?

stlaz replied:
`waitForConfigMapCAContent()` appears to be fetching the cluster authn CM first to compare it with the expectation:
https://github.com/kubernetes/kubernetes/blob/4044f71121e5c51fa5cb12f6f98fe93c118f7745/test/integration/apiserver/certreload/certreload_test.go#L312-L329

I'll have a further look though, while I can't find it just now, it seems a bit odd we wouldn't wire the ClientCA and RequestHeaderCA dynamic certificate Notifier events to our queue.

stlaz replied:
This was a proactive change

liggitt replied:
OK, let's drop these changes to make this PR as small as possible and make sure that existing tests keep passing as is. I'm fine with a follow up to tweak these if we think that's necessary to make the test more robust.

stlaz replied:
You're right, now that I think about it we did not touch anything with regards to how the CA files are handled, a change in behavior here would be weird.

stlaz commented on pkg/controlplane/controller/clusterauthenticationtrust/cluster_authentication_trust_controller.go:
```diff
@@ -256,32 +253,25 @@ func getConfigMapDataFor(authenticationInfo ClusterAuthenticationInfo) (map[stri
+	if utilfeature.DefaultFeatureGate.Enabled(features.RemoteRequestHeaderUID) {
+		sliceWritesMapping["requestheader-uid-headers"] = authenticationInfo.RequestHeaderUIDHeaders.Value()
+	}
+	var err error
+	for key, configData := range sliceWritesMapping {
+		data[key], err = jsonSerializeStringSlice(configData)
 		// encoding errors aren't going to get better, so just fail on them.
-		data["requestheader-username-headers"], err = jsonSerializeStringSlice(authenticationInfo.RequestHeaderUsernameHeaders.Value())
-		if err != nil {
-			return nil, err
-		}
-		if utilfeature.DefaultFeatureGate.Enabled(features.RemoteRequestHeaderUID) && len(authenticationInfo.RequestHeaderUIDHeaders.Value()) > 0 {
-			data["requestheader-uid-headers"], err = jsonSerializeStringSlice(authenticationInfo.RequestHeaderUIDHeaders.Value())
-			if err != nil {
-				return nil, err
-			}
-		}
-
-		data["requestheader-group-headers"], err = jsonSerializeStringSlice(authenticationInfo.RequestHeaderGroupHeaders.Value())
-		if err != nil {
-			return nil, err
-		}
-		data["requestheader-extra-headers-prefix"], err = jsonSerializeStringSlice(authenticationInfo.RequestHeaderExtraHeaderPrefixes.Value())
-		if err != nil {
-			return nil, err
-		}
-
-		data["requestheader-client-ca-file"] = string(caBytes)
-		data["requestheader-allowed-names"], err = jsonSerializeStringSlice(authenticationInfo.RequestHeaderAllowedNames.Value())
```
sure

stlaz commented:
Pushed the minimal change + proper description in the commit message.

liggitt commented:
/lgtm
/approve

stlaz commented:
/test pull-kubernetes-unit
unrelated, though odd. Looks like a problem in DRA API?
```patch
    roundtrip.go:358: ResourceClaimTemplate: diff:   &resource.ResourceClaimTemplate{
          	TypeMeta:   {},
          	ObjectMeta: {Name: "鸚窷Wĸ敧Ŵ輄ʇ檶X", GenerateName: "穡A呑 »H¥ľğP脮Ĝ朻<ç", Namespace: "Ņɉ[ʯP蜚", SelfLink: "Ë鲴ȱ佩´so鹉", ...},
          	Spec: resource.ResourceClaimTemplateSpec{
          		ObjectMeta: {Name: "漍(2遴覬擺", GenerateName: "Ņn0ɪ", Namespace: "áPȾ忇)", UID: "桭D闊ǳ@這ŀ咣jdĂ水磑揑狼", ...},
          		Spec: resource.ResourceClaimSpec{
          			Devices: resource.DeviceClaim{
          				Requests: []resource.DeviceRequest{
          					{
          						... // 4 identical fields
          						Count:       2215404912529444782,
          						AdminAccess: &false,
          						FirstAvailable: []resource.DeviceSubRequest{
          							{
          								Name:            "鸮棤$ɷɕ",
          								DeviceClassName: "Ȍ桏Ǐ蓉惻Jp蒥",
          								Selectors:       nil,
        - 								AllocationMode:  "",
        + 								AllocationMode:  "ExactCount",
          								Count:           -6163363086737291551,
          							},
          						},
          					},
          				},
```

liggitt commented:
> unrelated, though odd. Looks like a problem in DRA API?
> 
> ```diff
>     roundtrip.go:358: ResourceClaimTemplate: diff:   &resource.ResourceClaimTemplate{
>           	TypeMeta:   {},
>           	ObjectMeta: {Name: "鸚窷Wĸ敧Ŵ輄ʇ檶X", GenerateName: "穡A呑 »H¥ľğP脮Ĝ朻<ç", Namespace: "Ņɉ[ʯP蜚", SelfLink: "Ë鲴ȱ佩´so鹉", ...},
>           	Spec: resource.ResourceClaimTemplateSpec{
>           		ObjectMeta: {Name: "漍(2遴覬擺", GenerateName: "Ņn0ɪ", Namespace: "áPȾ忇)", UID: "桭D闊ǳ@這ŀ咣jdĂ水磑揑狼", ...},
>           		Spec: resource.ResourceClaimSpec{
>           			Devices: resource.DeviceClaim{
>           				Requests: []resource.DeviceRequest{
>           					{
>           						... // 4 identical fields
>           						Count:       2215404912529444782,
>           						AdminAccess: &false,
>           						FirstAvailable: []resource.DeviceSubRequest{
>           							{
>           								Name:            "鸮棤$ɷɕ",
>           								DeviceClassName: "Ȍ桏Ǐ蓉惻Jp蒥",
>           								Selectors:       nil,
>         - 								AllocationMode:  "",
>         + 								AllocationMode:  "ExactCount",
>           								Count:           -6163363086737291551,
>           							},
>           						},
>           					},
>           				},
> ```

@johnbelamaric @pohly issue in the DRA fuzzing setup?

stlaz commented:
/test pull-kubernetes-integration
looks unrelated, looks like etcd was still running when the test tried to launch another instance?
```
{"level":"fatal","ts":"2025-03-10T14:45:11.545692Z","caller":"etcdmain/etcd.go:204",
"msg":"discovery failed","error":"listen tcp 127.0.0.1:43397: bind: address already in use",
"stacktrace":"go.etcd.io/etcd/server/v3/etcdmain.startEtcdOrProxyV2\n\tgo.etcd.io/etcd/server/v3/etcdmain/etcd.go:204\ngo.etcd.io/etcd/server/v3/etcdmain.Main\n\tgo.etcd.io/etcd/server/v3/etcdmain/main.go:40\nmain.main\n\tgo.etcd.io/etcd/server/v3/main.go:31\nruntime.main\n\truntime/proc.go:271"}

    etcd.go:260: unable to start etcd: could not start etcd
--- FAIL: TestSchedulerPerf/EventHandlingPodUpdate/1Node_10Pods (35.09s)
```

stlaz commented:
The follow-up that uses this PR for UID headers can now be found at https://github.com/kubernetes/kubernetes/pull/130560
